### DISCUSSION TITLE
ЛР 4 для ИВТ. Вопросы по реализации Data Memory

### DISCUSSION THREAD

**Discussion by BROsandr** [MPSU/exAPS#34] [Вопросы и ответы] [answered]
> При этом она должна иметь видимость ограничения доступных адресов 0xNN000000 – 0xNN0003FC, где NN это последние две цифры студенческого. Например, если мой студенческий 001905, то 256 слов памяти должны быть расположены по адресам 0x05000000 – 0x05003FC

0xNN0003FC - 0xNN000000 = 0x000003FC = 1020 <br>
1020 / 4 = 255 слов <br>
Если захватывать граничные байты, то ( 1020 + 1 ) / 4 = 255.25 слов

**Comment by HepoH3** [comment, marked as answer]
По адресу 3FC размещен нулевой байт 255-го слова (256-го, если считать с единицы):

```
0x000 <- нулевой байт нулевого слова
0x003 <- третий байт нулевого слова
0x004 <- нулевой байт первого слова
0x007 <- третий байт первого слова
...
0x3FC <- нулевой байт 255-го слова
```

Поскольку в памяти инструкций данные хранятся словами, а не байтами (`reg [31:0] Memory [0:255]`), после преобразований адреса 0x3fc>>2=0xFF в память придет адрес 255го слова, и она вернет его целиком.
Если на лабе вы делали память инструкций как `reg [7:0] mem [0:1023]`, то вам не надо делать сдвиг, но адрес у вас должен быть уже 10битный.

**Reply by mrz25** [reply]
почему десятибитный, если по идее на вход Instruction Memory подается 32 бита адреса, из которых мы берем [7:0], то есть 8-битный адрес, а на DataMemory подается 32 бита адреса и там мы берем [9:2]?

**Reply by mrz25** [reply]
мы же эти 10 бит значащих урезаем до 8

**Reply by HepoH3** [reply]
data memory пока тут не трогается, в instruction memory приходит 32битный адрес, который потом сокращается до 8 бит значащих бит (в случае, если память делалась как `reg [31:0] mem [0:255]`) или 10 значащих бит (в случае, если память делалась как `reg [7:0] mem [0:1023]`).
Если скажете как у вас реализована память инструкций, будет проще дальше этот момент обсуждать, чтобы не прыгать между вариантами (просто в методичках сказано делать как `reg[31:0]...`, но на лабах я сразу предлагал делать `reg [7:0]...`, поэтому есть некоторое разночтение).

**Reply by mrz25** [reply]
вот я делал как вы на лабах говорили, то есть вот так выглядит моя память в Instruction Memory
    reg [7:0] RAM [0:1023];
и соответственно я не понимаю откуда берутся 10 бит, потому что адрес 8-битный в RAM....

**Reply by HepoH3** [reply]
> вот я делал как вы на лабах говорили, то есть вот так выглядит моя память в Instruction Memory reg [7:0] RAM [0:1023]; и соответственно я не понимаю откуда берутся 10 бит, потому что адрес 8-битный в RAM....

если делали 8битную память (`reg[7:0]`) то просто сделайте 10битный адрес. Вы же увеличили число элементов до 1024 (а раньше на лабе наверно делали 256). Раз элементов в памяти больше, конечно же вам нужно нарастить разрядность адреса.

**Reply by mrz25** [reply]
точно, сразу не догнал, но тогда получается, что и в DataMemory адрес нужно делать 10-битным? Но ведь мы там берем от 32-битного адреса [9:2], то есть к этим 8 битам нужно слева два нуля дописать?

**Reply by HepoH3** [reply]
Для памяти данных все то же самое: если она у вас восьмибитная, то адрес должен быть 10битным без всяких отбрасываний (просто берете младшие 10 бит).

**Reply by HepoH3** [reply]
Можно взять биты [9:2] и добить слева нулями, но проще конечно просто взять [9:0].

**Reply by mrz25** [reply]
так, с этим понял, не понял как можно приравнивать то что мы берем [9:0] и 00[9:2], если условно адрес значащий будет 3FB (11 1111 1011), то есть эти две единицы исчезнут во 2 случае

**Reply by HepoH3** [reply]
> так, с этим понял, не понял как можно приравнивать то что мы берем [9:0] и 00[9:2], если условно адрес значащий будет 3FB (11 1111 1011), то есть эти две единицы исчезнут во 2 случае

В данной реализации не может быть адреса 3fb. Там в начале оговорка, что в этой лабе пока делаем только поддержку стягивания слов (инструкции lh, lhu, lb, lbu не поддерживаются). Слова в памяти с выровненным доступом (а у нас такая) могут начинаться только с адресов, кратных четырем.

**Reply by HepoH3** [reply]
@mrz25, @BROsandr — ну как, вопросы остались или разобрались?

**Reply by mrz25** [reply]
> @mrz25, @BROsandr — ну как, вопросы остались или разобрались?

вы знаете, вроде со всем разобрались, написал программу бинарную для вычисления функции своего варианта, все прекрасно работает, потому что есть зацикливание PC = PC + 0, но, вбивая, в ассемблер свою функцию столкнулся с тем, что не понимаю как нужно его правильно зациклить, то есть выход из программы непонятно как сделать

**Reply by mrz25** [reply]
![image](https://user-images.githubusercontent.com/71031775/198854275-642b4b8d-4c05-4d23-a7be-88e3e2c346ce.png)
то есть вот, com_32 - конечная точка, при переходе на которую выдает правильный результат и если там пусто, то в регистрах нужные значения и все прекрасно, но если, условно дописать там "j com_32" с типо зацикливанием, то регистры ломаются и там появляются неверные значения, если в мемори файл вбить значения, скажем так, без com_32 и зацикливания, то у меня PC сбивается и почему-то плюсует в один момент +6, а не +4 и все тоже идет по одному месту

update: неправильно задавал константу, теперь все прекрасно работает с шестнадцатиричным кодом с ассемблера, лаба сделана
